torch/csrc/jit/passes/peephole.cpp

```diff
@@ -14,359 +16,405 @@ static bool mustBeEqual(const c10::optional<T>& a, const c10::optional<T>& b) {
+  // translation we would see redundant Gemm ops with sub-optimal inputs. This
+  // flag is exposed so that ONNX export can pass `true` to get the fused
+  // behavior, but normal JIT peephole optimization is left alone.
+  void run(Block* block) {
+    for (auto it = block->nodes().begin(); it != block->nodes().end(); ++it) {
+      auto* node = *it;
 
-    // XXX: remember that if you want to simplify an expression by combining
-    // multiple nodes into a different one, then you need to check that they all
-    // belong to the given block
-    if (node->matches(
-            "aten::expand(Tensor self, int[] size, *, bool implicit) -> Tensor",
-            /*const_inputs=*/attr::size)) {
-      // x.expand(x.size()) == x
-      if (auto input_type =
-              node->namedInput(attr::self)->type()->cast<TensorType>()) {
-        auto expanded_sizes = node->get<c10::List<int64_t>>(attr::size);
-        auto input_type_sizes = input_type->sizes().concrete_sizes();
-        if (expanded_sizes.has_value() && input_type_sizes &&
-            c10::impl::toVector(*expanded_sizes) == *input_type_sizes) {
-          GRAPH_UPDATE(
-              *node,
-              " (x.expand(x.size()) == x) is replaced with ",
-              node->namedInput(attr::self)->debugName());
-          node->output()->replaceAllUsesWith(node->namedInput(attr::self));
-        }
-      }
-    } else if (node->matches("aten::t(Tensor self) -> Tensor")) {
-      // x.t().t() == x
-      Node* input_node = node->input()->node();
-      if (input_node->matches("aten::t(Tensor self) -> Tensor")) {
-        GRAPH_UPDATE(
-            *node,
-            " (x.t().t() == x) is replaced with ",
-            input_node->input()->debugName());
-        node->output()->replaceAllUsesWith(input_node->input());
+      for (Block* sub_block : node->blocks()) {
+        run(sub_block);
       }
-    } else if (node->matches(
-                   "aten::type_as(Tensor self, Tensor other) -> Tensor")) {
-      // x.type_as(y) == x iff x.type() == y.type()
-      auto self_type = node->input(0)->type()->expect<TensorType>();
-      auto other_type = node->input(1)->type()->expect<TensorType>();
-      if (mustBeEqual(self_type->scalarType(), other_type->scalarType()) &&
-          mustBeEqual(self_type->device(), other_type->device())) {
-        GRAPH_UPDATE(
-            *node,
-            " (x.type_as(y) == x) is replaced with ",
-            node->input(0)->debugName());
-        node->output()->replaceAllUsesWith(node->input(0));
+
+      // if either the inputs or outputs of an op alias graph's inputs or
+      // outputs, the transformations below are invalid
+      // An example:
+      //
+      // def test_write(x):
+      //     s = 0
+      //     s += x
+      //     s += x
+      //     return s
+      //
+      // any transformations that are still valid despite
+      // should be pulled above this statement
+
+      if (aliasDb_->mayContainAlias(
```

Comment:
Oh, you also need the `hasWriters` check you had before. Example:
```
def foo(x):
    z = x.clone()
    y = z + 0
    z.add_(...)
    return y + y
```
In this case, `y` and `z` do not directly alias any output/inputs, but it is not correct to optimize out the add.